### PR TITLE
salmon: init at 1.10.2

### LINKED ISSUE #252992: Package request: Salmon
Salmon is a bioinformatics tool to produce a highly accurate, transcript level quantification estimates from short-read RNA-seq data.
It's mostly written in C++. There's a similar tool already in nixpkgs (Kallisto), but they both have a few unique features, so packaging Salmon is still useful. It's also quite a popular tool.

* homepage URL: https://salmon.readthedocs.io/en/latest/index.html
* source URL: https://github.com/COMBINE-lab/salmon
* license: GLP v3.0
* platforms: linux, others?

### PR DESCRIPTION
**Pull request by Kupac**
## Description of changes

Add new package: Salmon

Resolves: https://github.com/NixOS/nixpkgs/issues/252992

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jfvillablanca** [comment]
Can you use the new [by-name](https://github.com/NixOS/nixpkgs/blob/5f3699998b3282c1aa5ae73834ad494bc237345c/pkgs/by-name/README.md) convention

**Comment by Kupac** [comment]
Thanks for sharing the link to the docs. I wanted to do this after seeing the talk at NixCon. But couldn't find the description. Ready for review now.

File: pkgs/by-name/sa/salmon/package.nix [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      rev = "salmon-v${version}";
+      sha256 = "sha256-JKbUFBEsqnENl4vFqve1FCd4TI3n9bRi2RNHC8QGQGc=";
+  };
+
+  src = fetchFromGitHub {
+          owner = "COMBINE-lab";
+          repo = "salmon";
+          rev = "v${version}";
+          sha256 = "sha256-kwqoUmVCqjr/xRxJjQKaFjjCQW+MFASHJ2f9OiAumNU=";
+        };
```
Comment (FliegendeWurst) [inline pkgs/by-name/sa/salmon/package.nix:40]:
```suggestion
  pufferFishSrc = fetchFromGitHub {
    owner = "COMBINE-lab";
    repo = "pufferfish";
    rev = "salmon-v${version}";
    sha256 = "sha256-JKbUFBEsqnENl4vFqve1FCd4TI3n9bRi2RNHC8QGQGc=";
  };

  src = fetchFromGitHub {
    owner = "COMBINE-lab";
    repo = "salmon";
    rev = "v${version}";
    sha256 = "sha256-kwqoUmVCqjr/xRxJjQKaFjjCQW+MFASHJ2f9OiAumNU=";
  };
```

File: pkgs/by-name/sa/salmon/package.nix [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    lbzip2
+    xz
+    bzip2
+    cereal_1_3_2
+    tbb_2021_8
+    libgff
+    libstaden-read
+  ] ++ lib.optionals stdenv.isDarwin [ libiconv ];
+
+  nativeBuildInputs = [ cmake boost pkg-config bash curl jemalloc ];
```
Comment (FliegendeWurst) [inline pkgs/by-name/sa/salmon/package.nix:61]:
```suggestion
    libstaden-read
    jemalloc
    boost
    curl
  ] ++ lib.optionals stdenv.isDarwin [ libiconv ];

  nativeBuildInputs = [ cmake pkg-config ];

  strictDeps = true;
```

File: pkgs/by-name/sa/salmon/package.nix [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+  meta = {
+    description = "A tool for quantifying the expression of transcripts " +
```
Comment (FliegendeWurst) [inline pkgs/by-name/sa/salmon/package.nix:64]:
```suggestion
    description = "Tool for quantifying the expression of transcripts " +
```

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

**Comment by Kupac** [comment]
Thanks for your time and for picking up this PR, Flying Sausage!

File: pkgs/by-name/sa/salmon/package.nix (hunk 2 above)
Comment (Kupac) [reply pkgs/by-name/sa/salmon/package.nix]: So you moved some stuff from native to regular build inputs, and then set strictDeps to true? What is this sorcery? Do you have any recommended reading for me?

Comment (FliegendeWurst) [reply pkgs/by-name/sa/salmon/package.nix]: > What is this sorcery? Do you have any recommended reading for me?

The difference between nativeBuildInputs and buildInputs is that the first set of packages is used at build time (when compiling) whereas the second set is required by the program that is being compiled. Enabling `strictDeps` enforces this separation a bit more than the default (lenient) settings. This is explained in more detail in the [docs](https://ryantm.github.io/nixpkgs/stdenv/stdenv/#ssec-stdenv-dependencies-overview).

**Comment by Kupac** [comment]
@FliegendeWurst looks like borg is stuck on something. I remember it took some time last time around, but not this long. Is this a problem?

**Comment by FliegendeWurst** [comment]
> @FliegendeWurst looks like borg is stuck on something. I remember it took some time last time around, but not this long. Is this a problem?  

Yeah ARM Mac has been stuck the last few days. Not really a problem for this package.

File: pkgs/by-name/li/libgff/package.nix [hunk 4]
```diff
@@ -0,0 +1,35 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+}:
+
+stdenv.mkDerivation rec {
+  pname = "libgff";
+  version = "2.0.0";
+  src = fetchFromGitHub {
```
Comment (ambroisie) [inline pkgs/by-name/li/libgff/package.nix:10]:
```suggestion

  src = fetchFromGitHub {
```

File: pkgs/by-name/li/libgff/package.nix [hunk 5]
```diff
@@ -0,0 +1,35 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+}:
+
+stdenv.mkDerivation rec {
```
Comment (ambroisie) [inline pkgs/by-name/li/libgff/package.nix:7]: You should use the new [`finalAttrs`](https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes) idiom.

File: pkgs/by-name/li/libgff/package.nix [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  pname = "libgff";
+  version = "2.0.0";
+  src = fetchFromGitHub {
+    owner = "COMBINE-lab";
+    repo = "libgff";
+    rev = "v${version}";
+    sha256 = "sha256-ZCb3UyuB/+ykrYFQ9E5VytT65gAAULiOzIEu5IXISTc=";
+  };
+  nativeBuildInputs = [ cmake ];
```
Comment (ambroisie) [inline pkgs/by-name/li/libgff/package.nix:16]:
```suggestion

  nativeBuildInputs = [ cmake ];
```

File: pkgs/by-name/li/libgff/package.nix [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    description = "A lightweight GTF/GFF parsers exposing a C++ interface";
+    longDescription = ''
+    This is an attempt to perform a simple "libraryfication" of the GFF/GTF
+    parsing code that is used in GFFRead codebase. There are not many (any?)
+    relatively lightweight GTF/GFF parsers exposing a C++ interface, and the
+    goal of this library is to provide this functionality without the necessity
+    of drawing in a heavy-weight dependency like SeqAn. Note: This library
+    draws directly from the code in GFFRead and GCLib, and exists primarily to
+    remove functionality (and hence code) that is unnecessary for our
+    downstream purposes. '';
```
Comment (ambroisie) [inline pkgs/by-name/li/libgff/package.nix:28]: I probably wouldn't even add the `longDescription`, but please at least indent it.

File: pkgs/by-name/li/libstaden-read/package.nix [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+, bzip2
+, xz
+, zlib
+}:
+
+stdenv.mkDerivation rec {
+  # Same name as the Debian library
+  pname = "libstaden-read";
+  version = "1.15.0";
+  src = fetchFromGitHub {
```
Comment (ambroisie) [inline pkgs/by-name/li/libstaden-read/package.nix:15]:
```suggestion

  src = fetchFromGitHub {
```

File: pkgs/by-name/li/libstaden-read/package.nix [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  '';
+
+  meta = {
+    description = "C library for reading/writing various DNA sequence formats";
+    longDescription = ''
+    Libstaden-read is a library of file reading and writing code to provide a
+    general purpose SAM/BAM/CRAM, trace file (and Experiment File) reading
+    interface. Programmatically {S,B,CR}AM can be manipulated using the
+    scram_*() API functions while DNA Chromatogram ("trace") files can be read
+    using the read_reading() function.'';
```
Comment (ambroisie) [inline pkgs/by-name/li/libstaden-read/package.nix:45]: Same comment about `longDescription`.

File: pkgs/by-name/li/libstaden-read/package.nix [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    homepage = "https://staden.sourceforge.net";
+    downloadPage = "https://github.com/jkbonfield/io_lib/releases";
```
Comment (ambroisie) [inline pkgs/by-name/li/libstaden-read/package.nix:47]: TIL about [`downloadPage`](https://nixos.org/manual/nixpkgs/unstable/#var-meta-downloadPage). You should also define `changelog` which is more popular (assuming you use `finalAttrs`).

```suggestion
    downloadPage = "https://github.com/jkbonfield/io_lib/releases";
    changelog = "https://github.com/jkbonfield/io_lib/blob/${finalAttrs.src.rev}/CHANGES";
```

File: pkgs/by-name/li/libstaden-read/package.nix [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+, stdenv
+, autoreconfHook
+, fetchFromGitHub
+, gcc
+, bzip2
+, xz
+, zlib
+}:
+
+stdenv.mkDerivation rec {
```
Comment (ambroisie) [inline pkgs/by-name/li/libstaden-read/package.nix:11]: Same comment about `finalAttrs`.

File: pkgs/by-name/sa/salmon/package.nix [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+, libiconv
+, libstaden-read
+, lzlib
+, pkg-config
+, tbb_2021_8
+, xz
+, zlib
+}:
+
+stdenv.mkDerivation rec {
```
Comment (ambroisie) [inline pkgs/by-name/sa/salmon/package.nix:24]: Same comment about `finalAttrs`.

File: pkgs/by-name/sa/salmon/package.nix [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    curl
+  ] ++ lib.optionals stdenv.isDarwin [ libiconv ];
+
+  nativeBuildInputs = [ cmake pkg-config ];
+
+  strictDeps = true;
+
+  meta = {
+    description = "Tool for quantifying the expression of transcripts " +
+                  "using RNA-seq data";
```
Comment (ambroisie) [inline pkgs/by-name/sa/salmon/package.nix:70]: Just one line please.

```suggestion
    description = "Tool for quantifying the expression of transcripts using RNA-seq data";
```

File: pkgs/by-name/sa/salmon/package.nix [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+      using RNA-seq data. Salmon uses new algorithms (specifically,
+      coupling the concept of quasi-mapping with a two-phase inference
+      procedure) to provide accurate expression estimates very quickly
+      and while using little memory. Salmon performs its inference using
+      an expressive and realistic model of RNA-seq data that takes into
+      account experimental attributes and biases commonly observed in
+      real RNA-seq data.
+    '';
+    homepage = "https://combine-lab.github.io/salmon";
+    downloadPage = "https://github.com/COMBINE-lab/salmon/releases";
```
Comment (ambroisie) [inline pkgs/by-name/sa/salmon/package.nix:82]:
```suggestion
    downloadPage = "https://github.com/COMBINE-lab/salmon/releases";
    changelog = "https://github.com/COMBINE-lab/salmon/releases/tag/v${finalAttrs.version}";
```

**Review by ambroisie — CHANGES_REQUESTED** [review]

File: pkgs/by-name/sa/salmon/package.nix (hunk 13 above)
Comment (Kupac) [reply pkgs/by-name/sa/salmon/package.nix]: Yeh, I'm a bit obsessed with trying to keep it within 80 characters. I tried it in another way :)

File: pkgs/by-name/sa/salmon/package.nix [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    changelog = "https://github.com/COMBINE-lab/salmon/releases/tag/" +
+                "v${finalAttrs.version}";
```
Comment (ambroisie) [inline pkgs/by-name/sa/salmon/package.nix:84]: Same comment about not needing to wrap to 80 columns.

**Review by ambroisie — APPROVED** [review]
Formatting nit, but LGTM.

**Comment by Kupac** [comment]
What else can I do to get this merged? @ambroisie @FliegendeWurst

**Comment by ambroisie** [comment]
Unfortunately I don't have the merge bit.

So posting it to the "PRs already reviewed" thread on Discourse and/or waiting for a wandering merge-enabled-reviewer is the usual solution.

File: pkgs/by-name/sa/salmon/fetch-pufferfish.patch [hunk 16]
```diff
@@ -0,0 +1,66 @@
+diff --git a/scripts/fetchPufferfish.sh b/scripts/fetchPufferfish.sh
+index bf2574e0..42582806 100755
+--- a/scripts/fetchPufferfish.sh
++++ b/scripts/fetchPufferfish.sh
+@@ -1,4 +1,4 @@
+-#!/bin/bash
++#!${bash}/bin/bash
```
Comment (jbedo) [inline pkgs/by-name/sa/salmon/fetch-pufferfish.patch:7]: This doesn't seem right: this is not a nix expression and the `bash` environment variable isn't set, so this likely has no effect.

File: pkgs/by-name/li/libgff/package.nix [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+    rev = "v${finalAttrs.version}";
+    hash = "sha256-ZCb3UyuB/+ykrYFQ9E5VytT65gAAULiOzIEu5IXISTc=";
+  };
+
+  nativeBuildInputs = [ cmake ];
+
+  meta = {
+    description = "A lightweight GTF/GFF parsers exposing a C++ interface";
+    homepage = "https://github.com/COMBINE-lab/libgff";
+    downloadPage = "https://github.com/COMBINE-lab/libgff/releases";
```
Comment (jbedo) [inline pkgs/by-name/li/libgff/package.nix:23]: may as well set `changelog` like salmon

**Review by jbedo — CHANGES_REQUESTED** [review]
Tested fully functional on some simulated reads.

**Comment by jbedo** [comment]
Can you resolve the outstanding changes to unblock the merge?

**Comment by Kupac** [comment]
Ah, I missed the 2nd comment. Will do right away

**Comment by Kupac** [comment]
Should be OK now, pending ofborg eval

File: pkgs/by-name/sa/salmon/package.nix [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+, bash
+, boost
+, bzip2
+, cereal_1_3_2
+, cmake
+, curl
+, fetchFromGitHub
+, gcc
+, jemalloc
+, lbzip2
```
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/sa/salmon/package.nix:13]:
```suggestion
```
don't think this is required.

File: pkgs/by-name/sa/salmon/package.nix [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+    # Use pufferfish source fetched by nix
+    ./fetch-pufferfish.patch
+  ];
+
+  postPatch = "patchShebangs .";
+
+  buildInputs = [
+    zlib
+    lzlib
+    lbzip2
```
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/sa/salmon/package.nix:52]:
```suggestion
```

File: pkgs/by-name/sa/salmon/package.nix [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+, cmake
+, curl
+, fetchFromGitHub
+, gcc
+, jemalloc
+, lbzip2
+, libgff
+, libiconv
+, libstaden-read
+, lzlib
```
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/sa/salmon/package.nix:17]:
```suggestion
```

**Review by ghost — COMMENTED** [review]
a few minor comments.
- can remove gcc from the list of inputs
- can remove llzlib, lbzip2  (this should make darwin build)
- can revert change to all-packages.nix and do the boost override in the salmon package.

File: pkgs/by-name/sa/salmon/package.nix [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+    lzlib
+    lbzip2
+    xz
+    bzip2
+    cereal_1_3_2
+    tbb_2021_8
+    libgff
+    libstaden-read
+    jemalloc
+    boost
```
[the remark this answers is not in the document]
Comment (Kupac) [reply pkgs/by-name/sa/salmon/package.nix:59]: This is awesome, I didn't know it was legal to do it this way! :) Having the override in a separate location was a bit obscure.

**Review by ghost — APPROVED** [review]
Result of `nixpkgs-review pr 253179` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>libgff</li>
    <li>libstaden-read</li>
    <li>salmon</li>
  </ul>
</details>

---

Result of `nixpkgs-review pr 253179` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>libgff</li>
    <li>libstaden-read</li>
    <li>salmon</li>
  </ul>
</details>

**Comment by Kupac** [comment]
@jbedo I added all the improvements by a-n-n-a-l-e-e. I wasn't entirely sure about excluding bzip2 and lzma, but anyway, I only ever saw fq or fq.gz files, so it should be fine. Pls let me know if I forgot something still, and whether it's ready to be merged.

**Review by jbedo — APPROVED** [review]

**Comment by jbedo** [comment]
> I wasn't entirely sure about excluding bzip2 and lzma, but anyway, I only ever saw fq or fq.gz files, so it should be fine. 

I agree, I've only seen gzipped fqs in the wild.

**Pull request merged**